pkg/scripts/install.go

```diff
@@ -13,177 +15,234 @@ import (
-	result := make(map[string]*downloadResult, len(scripts))
+// ScriptSourceManager manages various script handlers.
+type ScriptSourceManager interface {
+	Add(ScriptHandler)
+	SetDownloader(Downloader)
+	Installer
+}
 
-	for _, script := range scripts {
-		result[script] = &downloadResult{location: scriptsURL}
-		url, err := url.Parse(scriptsURL + "/" + script)
-		if err != nil {
-			result[script].err = err
-			continue
-		}
-		_, result[script].err = i.downloader.Download(url, filepath.Join(dstDir, script))
-	}
+// DefaultScriptSourceManager manages the default script lookup and installation
+// for source-to-image.
+type DefaultScriptSourceManager struct {
+	Image      string
+	ScriptsURL string
+	download   Downloader
+	docker     docker.Docker
+	dockerAuth dockerClient.AuthConfiguration
+	sources    *list.List
+	fs         util.FileSystem
+}
 
-	return result
+// Add registers a new script source handler.
+func (m *DefaultScriptSourceManager) Add(s ScriptHandler) {
+	if m.sources == nil {
+		m.sources = list.New()
+	}
+	m.sources.PushBack(s)
 }
 
-func (i *installer) install(scripts []string, userResults, sourceResults, defaultResults map[string]*downloadResult, dstDir string) []api.InstallResult {
-	resultList := make([]api.InstallResult, len(scripts))
+// NewInstaller returns a new instance of the default Installer implementation
+func NewInstaller(image string, scriptsURL string, docker docker.Docker, auth dockerClient.AuthConfiguration) Installer {
+	m := DefaultScriptSourceManager{
+		Image:      image,
+		ScriptsURL: scriptsURL,
+		dockerAuth: auth,
+		docker:     docker,
+		fs:         util.NewFileSystem(),
+		download:   NewDownloader(),
+	}
+	// Order is important here, first we try to get the scripts from provided URL,
+	// then we look into sources and check for .s2i/bin scripts.
+	if len(m.ScriptsURL) > 0 {
+		m.Add(&UrlScriptHandler{URL: m.ScriptsURL, download: m.download, fs: m.fs})
+	}
+
+	m.Add(&SourceScriptHandler{fs: m.fs})
 
-	locationsResultsMap := map[string]map[string]*downloadResult{
-		api.UserScripts:    userResults,
-		api.SourceScripts:  sourceResults,
-		api.DefaultScripts: defaultResults,
+	// If the detection handlers above fail, try to get the script url from the
+	// docker image itself.
+	defaultURL, err := m.docker.GetScriptsURL(m.Image)
+	if err == nil && defaultURL != "" {
+		m.Add(&UrlScriptHandler{URL: defaultURL, download: m.download, fs: m.fs})
 	}
+	return &m
+}
 
-	// iterate over scripts
-	for idx, script := range scripts {
-		result := api.InstallResult{Script: script}
+// InstallRequired Downloads and installs required scripts into dstDir, the result is a
+// map of scripts with detailed information about each of the scripts install process
+// with error if installing some of them failed
+func (i *DefaultScriptSourceManager) InstallRequired(scripts []string, dstDir string) ([]api.InstallResult, error) {
+	result := i.InstallOptional(scripts, dstDir)
```

Comment:
I prefer the previous construct where both InstallRequired and InstallOptional called private run method. Seeing InstallRequired calling InstallOptional seems odd to me.